src/main/java/io/cdap/plugin/gcp/dataplex/source/DataplexBatchSource.java

```diff
@@ -0,0 +1,522 @@
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+
+/**
+ * Batch Source that reads data from Dataplex entities (BigQuery or Cloud Storage).
+ * <p>
+ * {@code Object} is the first parameter because that is the key used
+ * by Hadoop's {@code TextInputFormat}.
+ * Object is the second parameter because that is the value used
+ * by Hadoop's {@code TextInputFormat}.
+ * {@code StructuredRecord} is the third parameter because that is the output given by
+ * Hadoop's {@code TextInputFormat}. All the plugins included with Hydrator operate on
+ * StructuredRecord.
+ */
+@Plugin(type = BatchSource.PLUGIN_TYPE)
+@Name(DataplexBatchSource.NAME)
+@Description("Dataplex Source")
+public class DataplexBatchSource extends BatchSource<Object, Object, StructuredRecord> {
+  public static final String NAME = "Dataplex";
+  private static final String BQ_TEMP_BUCKET_NAME_PREFIX = "dataplex-bq-source-bucket-";
+  private static final String BQ_TEMP_BUCKET_NAME_TEMPLATE = BQ_TEMP_BUCKET_NAME_PREFIX + "%s";
+  private static final String CONFIG_TEMPORARY_TABLE_NAME = "cdap.bq.source.temporary.table.name";
+  private static final String GCS_TEMP_BUCKET_NAME = "dataplex-cdf-" + UUID.randomUUID();
+  private static final String DATAPLEX_TASK_ARGS = "TASK_ARGS";
+
+  private static final Logger LOG = LoggerFactory.getLogger(DataplexBatchSource.class);
+  private static Entity entity;
+  private static String dataset;
+  private static String datasetProject;
+  private static Schema outputSchema;
+  private static String tableId;
+  private final BigQueryAvroToStructuredTransformer transformer = new BigQueryAvroToStructuredTransformer();
+  private final DataplexBatchSourceConfig config;
+  private Configuration configuration;
+  private String bucketPath;
+
+  public DataplexBatchSource(DataplexBatchSourceConfig dataplexBatchSourceConfig) {
+    this.config = dataplexBatchSourceConfig;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    super.configurePipeline(pipelineConfigurer);
+    StageConfigurer configurer = pipelineConfigurer.getStageConfigurer();
+    FailureCollector collector = configurer.getFailureCollector();
+    if (!config.getConnection().canConnect() || config.getServiceAccountType() == null ||
+      (config.isServiceAccountFilePath() && config.autoServiceAccountUnavailable()) ||
+      (config.tryGetProject() == null)) {
+      // ValidatingInputFormat plugin setup is mandatory.Otherwise pipeline will fail at runtime in case of GCS entity.
+      config.setupValidatingInputFormat(pipelineConfigurer, collector, null);
+      return;
+    }
+    config.validateServiceAccount(collector);
```

Comment:
Instead of calling every config validation method individually, can we create `validate(failureCollector)` method in config class which internally calls all the config validation methods?
Same `validate(failureCollector)` method can be called in `prepareRun` as well

Reply:
merged the validation methods together for BQ entities but same can not be done for GCS entities as in configurePipeline we setValidatingInputFormat and in prepareRun we getValidatingInputFormat as set at design time.